extension-workspace-references.md

```diff
@@ -0,0 +1,182 @@
+# workspace/xreferences extension to LSP
+
+The `workspace/xreferences` extension to the Language Server Protocol (LSP) enables a language server to export all of the references made from the workspaces code to its dependencies. Additionally, a new `textDocument/xdefinition` method performs the equivalent of `textDocument/definition` while returning metadata about the definition and _optionally_ a concrete location to it.
+
+Use case: clients of a language server can invoke `workspace/xreferences` in order to find references to dependencies. This information can then be stored inside of a database, which allows the caller to create a 'global mapping' of symbols in dependencies to the workspaces they are used in (e.g. to see "how do other people use this symbol?"). A user would perform `textDocument/xdefinition` in order to locate the metadata about the symbol they are interested in and find its references in the database.
+
+### Initialization
+
+`ServerCapabilities` may contain a new field to indicate server-side support for this extension:
+
+```typescript
+interface ServerCapabilities {
+  /* ... all fields from the base ServerCapabilities ... */
+
+  /**
+   * The server provides workspace references exporting support.
+   */
+  xworkspaceReferencesProvider?: boolean;
+}
+```
+
+#### Workspace References Request
+
+The workspace references request is sent from the client to the server to export project-wide references to dependencies. That is, the response only returns references in the project to symbols defined in dependencies.
+
+_Request_
+* method: 'workspace/xreferences'
+* params: `WorkspaceReferencesParams` defined as follows:
+```typescript
+/**
+ * The parameters of a Workspace References Request.
+ */
+interface WorkspaceReferencesParams {
+}
+```
+
+_Response_:
+* result: `ReferenceInformation[]` defined as follows:
+```typescript
+/**
+ * Represents information about a reference to programming constructs like
+ * variables, classes, interfaces etc.
+ */
+interface ReferenceInformation {
+    /**
+     * The location in the workspace where the `symbol` has been referenced.
+     */
+    reference: Location;
+
+    /**
+     * Metadata information describing the symbol being referenced.
+     */
+    symbol: SymbolDescriptor;
+}
+```
+* error: code and message set in case an exception happens during the workspace references request.
+
+Where `SymbolDescriptor` is defined as follows:
+
+```typescript
+/**
+ * Represents information about a programming construct like a variable, class,
+ * interface etc that has a reference to it. Effectively, it contains data similar
+ * to SymbolInformation except all fields are optional.
+ *
+ * SymbolDescriptor usually uniquely identifies a symbol, but it is
+ * not guaranteed to do so.
+ */
+interface SymbolDescriptor {
+    /**
+     * The name of this symbol (same as `SymbolInformation.name`).
+     */
+    name?: string;
```

Comment:
> What is `name` for a constructor?

I assume you mean when a reference to a constructor has been made (just making sure we're speaking in the same terms). In that situation, it would just be the name of the constructor. It's not really any different than what you would get as a response to `workspace/symbol` or `textDocument/symbol`.

> Will this information ever be shown to the user?

I _think_ that it would not be. I believe we would pass this information back into a new LSP method which would return an exact location -- and from that you would present whatever you want to the user (ex: hover). But it's no different than `workspace/symbol` output so sure, it could be displayed to the user.